File: php/binary_search.php
<?php
# 二分法
# 必须是有序的数列
function bin_search($arr = [], $value = '')
{
	$low = 0;
	$high = count($arr) - 1;
	$flag = (count($arr) % 2 == 0) ? 'Even' : 'Odd'; 
	$count = 0;
	while ($low <= $high) {
		$count++;
		if ($flag === 'Even') {
			$mid = floor(($low + $high) / 2);
		} 
		if ($flag === 'Odd') {
			$mid = ceil(($low + $high) / 2);		
		}

		$guess = $arr[$mid];
		
		if ($guess == $value) {
			return [
				'key' 	=> $mid,
				'value'	=> $guess,
				'times' => $count
			];
		} 
		if ($guess < $value) {
			$low = $mid;
		} 
		if ($guess > $value) {
			$high = $mid;
		}
	}
	
	return null;
}

$arr = [9, 8, 7, 6, 5, 4, 3, 2, 1];
$arr = array_reverse($arr);

$value = 9;
$ret = bin_search($arr, $value);
print_r($ret);

File: python/binary_search.py
# 搜索list中的某个数
def binnary_search(list, item):
	low = 0
	high = len(list) - 1
	count = 0
	while low <= high:
		count = count + 1 
		mid = low + high
		guess = list[mid]
		# 移动 low 和 high 的位置
		if guess == item:
			return mid, count
		elif guess > item:
			high = mid - 1
		elif guess < item:
			low = mid + 1
			
	return None
			
list = [1, 2, 3, 4, 5, 6, 7, 8, 9]
# item1 = -1
item2 = 3
# res1, count1 = binnary_search(list, item1)
res2, count2 = binnary_search(list, item2)
# print (res1, count1)
print (res2, count2)
